# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [ ]:
print('Escola de Dados da Alura')

Escola de Dados da Alura


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [ ]:
print('Nome: Jucenir\nSobrenome: Serafim')

Nome: Jucenir
Sobrenome: Serafim


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [ ]:
print('J\nU\nC\nE\nN\nI\nR')

J
U
C
E
N
I
R


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [ ]:
dia = 3
mes = 'junho'
ano = 1985

print(f'{dia} {mes} {ano} ')

3 junho 1985 


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
anoAtual = 2023

print(f'Ano atual: {anoAtual} ')

type(anoAtual)

Ano atual: 2023 


int

## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [ ]:
nome = input('Digite seu nome:\n')
print(f'Olá {nome}!')

Digite seu nome:
Pedro
Olá Pedro!


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [ ]:
nome = input('Digite seu nome:\n')
idade = int(input('Digite sua idade:\n'))


print(f'Olá, {nome}, você tem {idade}!')

Digite seu nome:
Juce
Digite sua idade:
38
Olá, Juce, você tem 38!


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [ ]:
nome = input('Digite seu nome:\n')
idade = int(input('Digite sua idade:\n'))
altura= float(input('Digite sua altura em metros\n'))

print(f'Olá, {nome}, você tem {idade} e mede {altura} metros!')

Digite seu nome:
Ju
Digite sua idade:
35
Digite sua altura em metros
1.85
Olá, Ju, você tem 35 e mede 1.85 metros!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
v1=int(input('Insira um valor:\n'))
v2=int(input('Insira outro valor:\n'))

soma=v1+v2

print(f'{v1}+{v2} é igual a {soma}')

Insira um valor:
5
Insira outro valor:
7
5+7 é igual a 12


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
v1=int(input('Insira um valor:\n'))
v2=int(input('Insira outro valor:\n'))
v3=int(input('Insira mais outro valor:\n'))
soma=v1+v2+v3

print(f'{v1} + {v2} + {v3} é igual a {soma}')

Insira um valor:
45
Insira outro valor:
45
Insira mais outro valor:
78
45 + 45 + 78 é igual a 168


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [ ]:
v1=int(input('Insira um valor:\n'))
v2=int(input('Insira outro valor:\n'))

sub=v1-v2

print(f'{v1} - {v2} é igual a {sub}')

Insira um valor:
5
Insira outro valor:
4
5 - 4 é igual a 1


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [ ]:
v1=int(input('Insira um valor:\n'))
v2=int(input('Insira outro valor:\n'))

prod=v1 * v2

print(f'{v1} * {v2} é igual a {prod}')

Insira um valor:
5
Insira outro valor:
4
5 * 4 é igual a 20


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
v1=int(input('Insira um valor para o numerador:\n'))
v2=int(input('Insira um valor maior que zero para o denominador:\n'))

divi=v1 / v2

print(f'{v1} / {v2} é igual a {divi}')

Insira um valor para o numerador:
2
Insira um valor maior que zero para o denominador:
2
2 / 2 é igual a 1.0


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
v1=int(input('Insira um valor para o operador:\n'))
v2=int(input('Insira um valor para a potência:\n'))

exponen=v1 ** v2

print(f'{v1} ** {v2} é igual a {exponen}')

Insira um valor para o operador:
3
Insira um valor para a potência:
3
3 ** 3 é igual a 27


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
v1=int(input('Insira um valor para o numerador:\n'))
v2=int(input('Insira um valor maior que zero para o denominador:\n'))

divi=v1 // v2

print(f'{v1} // {v2} é igual a {divi}')

Insira um valor para o numerador:
20
Insira um valor maior que zero para o denominador:
3
20 // 3 é igual a 6


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
v1=int(input('Insira um valor para o numerador:\n'))
v2=int(input('Insira um valor maior que zero para o denominador:\n'))

divi=v1 % v2

print(f'O resto da divisão de {v1} / {v2} é igual a {divi}')

Insira um valor para o numerador:
5
Insira um valor maior que zero para o denominador:
2
O resto da divisão de 5 / 2 é igual a 1


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [ ]:
n1=int(input('Insira a primeira nota do trimestre:\n'))
n2=int(input('Insira a segunda nota do trimestre:\n'))
n3=int(input('Insira a terceira nota do trimestre:\n'))

media = (n1+n2+n3)/3

print(f'Sua média neste trimestre é {media}')


Insira a primeira nota do trimestre:
70
Insira a segunda nota do trimestre:
80
Insira a terceira nota do trimestre:
90
Sua média neste trimestre é 80.0


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
n1=5
n2=12
n3=20
n4=15

p1=1
p2=2
p3=3
p4=4

media_ponderada=((n1*p1)+(n2*p2)+(n3*p3)+(n4*p4))/(p1+p2+p3+p4)

print(media_ponderada)

14.9


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [ ]:
frase='keep on fighting'
print(frase)

keep on fighting


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [ ]:
frase1=input('Escreva uma frase:\n')
print(frase1)

Escreva uma frase:
The future is bright
The future is bright


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [ ]:
texto = input('Escreva uma frase:\n')
capital_txt= texto.upper()
print(capital_txt)

Escreva uma frase:
stronger than yesterday
STRONGER THAN YESTERDAY


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [ ]:
texto = input('Escreva uma frase:\n')
low_txt= texto.lower()
print(low_txt)

Escreva uma frase:
BELIVE!
belive!


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [ ]:
frase= input('Escreva uma frase:\n')
no_space= frase.strip()
print(no_space)

Escreva uma frase:
        just do it!      
just do it!


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [ ]:
frase= input('Escreva uma frase:\n')
no_space= frase.strip()
print(no_space)

Escreva uma frase:
          gurl, really?    
gurl, really?


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [ ]:
frase= input('Escreva uma frase:\n')
no_space= frase.strip()
capital=no_space.upper()
print(capital)

Escreva uma frase:
      let's try it!    
LET'S TRY IT!


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [ ]:
frase = input('Escreva uma frase:\n')
sem_e = frase.replace('e', 'f')
print(sem_e)

Escreva uma frase:
elefante elevado
flffantf flfvado


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [ ]:
frase = input('Escreva uma frase:\n')
frase = frase.replace('a', '@')
print(frase)

Escreva uma frase:
alada amanda
@l@d@ @m@nd@


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [ ]:
frase = input('Escreva uma frase:\n')
frase = frase.replace('s', '$')
print(frase)

Escreva uma frase:
salário secundário
$alário $ecundário


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [ ]:
v1=int(input('Insira um número\n'))
v2=int(input('Insira outro número:\n'))

if v1>v2:
  print(f'{v1} é o número maior')
else:
  print(f'{v2} é o número maior')

Insira um número
15
Insira outro número:
20
20 é o número maior


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [ ]:
growth = int(input('Insira em número o crescimento da empresa esse mês:\n'))
if growth>0:
  print('porcentagem positiva')
else:
  print('porcentagem negativa')

Insira em número o crescimento da empresa esse mês:
20
porcentagem positiva


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
print('Vogal ou Consoante')
letter=input('Insira uma letra, o programa dirá se é vogal ou consoante:\n')
vowels=['a','e','i','o','u','A','E','I','O','U' ]

if letter in vowels:
  print(f'A letra {letter} é uma vogal')
else:
  print(f'A letra {letter} é uma consoante')

Vogal ou Consoante
Insira uma letra, o programa dirá se é vogal ou consoante:
B
A letra B é uma consoante


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:
car_prices = [80000, 75000, 70000, 68000, 72000, 78000, 76000, 65000, 70000, 75000,
              82000, 90000, 68000, 72000, 75000, 82000, 77000, 71000, 69000, 73000, 74000, 78000, 81000, 75000, 82000,
              83000, 85000, 88000, 92000, 96000, 69000, 71000, 68000, 72000, 75000, 79000]

car_max=max(car_prices)
car_min=min(car_prices)

print(f'Preço mais alto é de {car_max}\nPreço mais baixo é de {car_min}')

Preço mais alto é de 96000
Preço mais baixo é de 65000


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [ ]:
produtos=[]
prod1=float(input('Insira o preço do produto 1:\n'))
produtos.append(prod1)
prod2=float(input('Insira o preço do produto 2:\n'))
produtos.append(prod2)
prod3=float(input('Insira o preço do produto 3:\n'))
produtos.append(prod3)

prod_cheap=min(produtos)

print(f'Produto mais barato custa R$ {prod_cheap:.2f}')

Insira o preço do produto 1:
25.47
Insira o preço do produto 2:
45.80
Insira o preço do produto 3:
25.41
Produto mais barato custa R$ 25.41


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [ ]:
number=[]
num1=int(input('Insira um número inteiro:\n'))
number.append(num1)
num2=int(input('Insira outro número inteiro:\n'))
number.append(num2)
num3=int(input('Insira mais um número inteiro:\n'))
number.append(num3)

number.sort()

print(number)

Insira um número inteiro:
50
Insira outro número inteiro:
30
Insira mais um número inteiro:
10
[10, 30, 50]


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [ ]:
studying_time=input('Qual seu turno de estudo? (manhã, tarde, noite)\n')

if studying_time == 'manhã':
  print('Bom dia!')
elif studying_time == 'tarde':
  print('Boa tarde!')
elif studying_time == 'noite':
  print('Boa noite!')
else:
  print('Valor Inválido!')

Qual seu turno de estudo? (manhã, tarde, noite)
manhã
Bom dia!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [ ]:
number=int(input('Escreva um número\n'))
if number % 2 !=0:
  print('Número ímpar')
else:
  print('Número par')

Escreva um número
8
Número par


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [ ]:
number=input('Escreva um número\n')
if number.isdigit():
  print('Número Inteiro')
else:
  print('Número Decimal')


Escreva um número
5
Número Inteiro


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [ ]:
num1=int(input('Insira um número diferente de zero\n'))
num2=int(input('Insira outro número diferente de zero\n'))

ops=int(input('Qual operação deseja realizar?\nDigite 1 para adição;\nDigite 2 para subtração;\nDigite 3 para multiplicação;\nDigite 4 para divisão.\n'))
if ops == 1:
  soma = num1+num2
  print(f'A soma de {num1} + {num2} = {soma}')
elif ops == 2:
  sub = num1-num2
  print(f'A subtração de {num1} - {num2} = {sub}')
elif ops == 3:
  prod = num1*num2
  print(f'A multiplicação de {num1} * {num2} = {prod}')
elif ops == 4:
  divi = num1/num2
  print(f'A divisão de {num1} / {num2} = {divi}')

Insira um número diferente de zero
5
Insira outro número diferente de zero
2
Qual operação deseja realizar?
Digite 1 para adição;
Digite 2 para subtração;
Digite 3 para multiplicação;
Digite 4 para divisão.
4
A divisão de 5 / 2 = 2.5


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [ ]:
num1=int(input('Insira um número diferente de zero para representar um lado de um triângulo\n'))
num2=int(input('Insira outro número diferente de zero para representar um lado de um triângulo\n'))
num3=int(input('Insira mais outro número diferente de zero para representar um lado de um triângulo\n'))

if num1+num2 > num3 and num1+num3 > num2 and num2+num3 > num1:
    if num1==num2==num3:
      print('Triângulo Equilátero')
    elif num1==num2 or num1==num3 or num2==num3:
      print('Triângulo Isósceles')
    else:
      print('Triângulo Escaleno')
else:
  print('Não é um triângulo')



Insira um número diferente de zero para representar um lado de um triângulo
5
Insira outro número diferente de zero para representar um lado de um triângulo
7
Insira mais outro número diferente de zero para representar um lado de um triângulo
10
Triângulo Escaleno


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [ ]:
etanol=1.70
diesel=2.00

fuel=int(input('Qual combustível quer comprar?\nDigite 1 para ETANOL\nDigite 2 para DIESEL\n'))
liters= int(input('Insira a quantidade de litros em número inteiro que deseja abastecer:\n'))


if fuel == 1:
  if liters <= 15:
    price = (liters * etanol)
    discount = price * 0.02
    final_price = price - discount
    print(f'Você abasteceu R${price:.2f}\nVocê economizou R${discount:.2f}\nPreço Final R${final_price:.2f}')
  elif liters > 15:
    price = (liters * etanol)
    discount = price * 0.04
    final_price = price - discount
    print(f'Você abasteceu R${price:.2f}\nVocê economizou R${discount:.2f}\nPreço Final R${final_price:.2f}')

elif fuel == 2:
  if liters <= 15:
    price = (liters * etanol)
    discount = price * 0.03
    final_price = price - discount
    print(f'Você abasteceu R${price:.2f}\nVocê economizou R${discount:.2f}\nPreço Final R${final_price:.2f}')
  elif liters > 15:
    price = (liters * diesel)
    discount = price * 0.05
    final_price = price - discount
    print(f'Você abasteceu R${price:.2f}\nVocê economizou R${discount:.2f}\nPreço Final R${final_price:.2f}')


Qual combustível quer comprar?
Digite 1 para ETANOL
Digite 2 para DIESEL
2
Insira a quantidade de litros em número inteiro que deseja abastecer:
50
Você abasteceu R$100.00
Você economizou R$5.00
Preço Final R$95.00


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [ ]:

vendas_2022 = float(input("Digite o valor de vendas em 2022: "))
vendas_2023 = float(input("Digite o valor de vendas em 2023: "))

variacao_percentual = ((vendas_2023 - vendas_2022) / vendas_2022) * 100

print(f"Variação percentual de vendas: {variacao_percentual:.2f}%")

if variacao_percentual > 20:
    print("Sugestão: Bonificação para o time de vendas.")
elif 2 <= variacao_percentual <= 20:
    print("Sugestão: Pequena bonificação para o time de vendas.")
elif -10 <= variacao_percentual < 2:
    print("Sugestão: Planejamento de políticas de incentivo às vendas.")
else:
    print("Sugestão: Corte de gastos.")

Digite o valor de vendas em 2022: 5487.45
Digite o valor de vendas em 2023: 6587.74
Variação percentual de vendas: 20.05%
Sugestão: Bonificação para o time de vendas.


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [ ]:
n1= int(input('Insira um número inteiro:\n'))
n2= int(input('Insira um número inteiro maior que o primeiro inserido:\n'))

while n1 <= n2:
  print(n1)
  n1+=1


Insira um número inteiro:
10
Insira um número inteiro maior que o primeiro inserido:
20
10
11
12
13
14
15
16
17
18
19
20


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [ ]:
colA=4
colB=10

dias= 0

while colA < colB:
  print(f'Dias passados {dias}.\nNúmero de elementos colônia A {colA}.\nNúmero de elementos colônia B {colB}.')
  colA*=1.03
  colB*=1.015
  dias+=1
print(f'Levará {dias} dias para a colônia A ultrapassar ou igualar a colônia B.')

Dias passados 0.
Número de elementos colônia A 4.
Número de elementos colônia B 10.
Dias passados 1.
Número de elementos colônia A 4.12.
Número de elementos colônia B 10.149999999999999.
Dias passados 2.
Número de elementos colônia A 4.2436.
Número de elementos colônia B 10.302249999999997.
Dias passados 3.
Número de elementos colônia A 4.370908.
Número de elementos colônia B 10.456783749999996.
Dias passados 4.
Número de elementos colônia A 4.5020352400000005.
Número de elementos colônia B 10.613635506249995.
Dias passados 5.
Número de elementos colônia A 4.6370962972.
Número de elementos colônia B 10.772840038843745.
Dias passados 6.
Número de elementos colônia A 4.7762091861160005.
Número de elementos colônia B 10.934432639426399.
Dias passados 7.
Número de elementos colônia A 4.919495461699481.
Número de elementos colônia B 11.098449129017794.
Dias passados 8.
Número de elementos colônia A 5.0670803255504655.
Número de elementos colônia B 11.26492586595306.
Dias passados 9.
Número 

#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [ ]:
notas=[]


for _ in range(15):
  while True:
    nota=int(input('Insira a nota da avaliação (0 a 5):\n'))

    if 0<= nota <=5:
      notas.append(nota)
      break
    else:
      print('Nota inválida! Insira valores entre 0 e 5.')


Insira a nota da avaliação (0 a 5):
5
Insira a nota da avaliação (0 a 5):
7
Nota inválida! Insira valores entre 0 e 5.
Insira a nota da avaliação (0 a 5):
4
Insira a nota da avaliação (0 a 5):
3


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [ ]:
temperatures=[]

while True:
  addTemp=int(input('Adicionar temperatura (1 para sim e 2 para não):\n'))

  if addTemp == 1:
    temp=float(input('Insira uma temperatura em Celcius '))
    temperatures.append(temp)
  elif addTemp == 2:
    if not temperatures:
      print('Nenhuma temperatura foi adicionada')
    else:
      tempAverage=sum(temperatures)/len(temperatures)
      print(f'A média das temperaturas é: {tempAverage:.2f}°C')
    break
  else:
        print('Opção inválida. Use 1 para adicionar temperatura ou 2 para encerrar.')



Adicionar temperatura (1 para sim e 2 para não):
1
Insira uma temperatura em Celcius 40
Adicionar temperatura (1 para sim e 2 para não):
35
Opção inválida. Use 1 para adicionar temperatura ou 2 para encerrar.
Adicionar temperatura (1 para sim e 2 para não):
1
Insira uma temperatura em Celcius 35
Adicionar temperatura (1 para sim e 2 para não):
2
A média das temperaturas é: 37.50°C


#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [ ]:
import math

numb=int(input('Insira um número inteiro:\n'))

factorial_numb=math.factorial(numb)

print(f'O fatorial de {numb} é {factorial_numb}')

Insira um número inteiro:
5
O fatorial de 5 é 120


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [ ]:
multiplicando=int(input('TABUADA\nInsira um número interiro de 1 a 10: '))

if 1 <= multiplicando <= 10:
  print(f'Tabuada do {multiplicando}:')
  for multiplicador in range(1, 11): #o multiplicador não precisa existir previamente, é criado no loop for
    resultado=multiplicando * multiplicador
    print(f'{multiplicando} x {multiplicador} = {resultado}')
else:
    print('Por favor, insira um número inteiro de 1 a 10.')


TABUADA
Insira um número interiro de 1 a 10: 7
Tabuada do 7:
7 x 1 = 7
7 x 2 = 14
7 x 3 = 21
7 x 4 = 28
7 x 5 = 35
7 x 6 = 42
7 x 7 = 49
7 x 8 = 56
7 x 9 = 63
7 x 10 = 70


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

In [ ]:
numb=int(input('Insira um número que queira saber se é primo: '))
if numb <=1:
   print(f'{numb} não é um número primo')
else:
  is_prime = True

  for i in range(2, numb):
    if numb % i == 0:
      is_prime = False #caso haja divisor não é primo
      break

  if is_prime:
   print(f'{numb} é um número primo')
  else:
    print(f'{numb} não é um número primo')

Insira um número que queira saber se é primo: 10
10 não é um número primo


#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

In [ ]:
# Inicialize contadores para cada faixa etária
faixa_0_25 = 0
faixa_26_50 = 0
faixa_51_75 = 0
faixa_76_100 = 0

# Solicita ao usuário que insira as idades e calcula a distribuição
while True:
    idade = int(input("Insira a idade do cliente (ou um número negativo para encerrar): "))

    if idade < 0:
        break

    if idade >= 0 and idade <= 25:
        faixa_0_25 += 1
    elif idade >= 26 and idade <= 50:
        faixa_26_50 += 1
    elif idade >= 51 and idade <= 75:
        faixa_51_75 += 1
    elif idade >= 76 and idade <= 100:
        faixa_76_100 += 1

# Exibe a distribuição de idades
print("Distribuição de idades:")
print("[0-25]:", faixa_0_25)
print("[26-50]:", faixa_26_50)
print("[51-75]:", faixa_51_75)
print("[76-100]:", faixa_76_100)

Insira a idade do cliente (ou um número negativo para encerrar): 25
Insira a idade do cliente (ou um número negativo para encerrar): 38
Insira a idade do cliente (ou um número negativo para encerrar): 90
Insira a idade do cliente (ou um número negativo para encerrar): 45
Insira a idade do cliente (ou um número negativo para encerrar): 74
Insira a idade do cliente (ou um número negativo para encerrar): 8
Insira a idade do cliente (ou um número negativo para encerrar): 65
Insira a idade do cliente (ou um número negativo para encerrar): 10
Insira a idade do cliente (ou um número negativo para encerrar): -1
Distribuição de idades:
[0-25]: 3
[26-50]: 2
[51-75]: 2
[76-100]: 1


#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

In [ ]:
# Inicializamos as variáveis contadoras
votos_candidato1 = 0
votos_candidato2 = 0
votos_candidato3 = 0
votos_candidato4 = 0
votos_nulos = 0
votos_branco = 0

# Início do laço para ler os votos
for i in range(0,20):
    voto = int(input('Informe seu voto: '))
    if voto == 1:
        votos_candidato1 += 1
    elif voto == 2:
        votos_candidato2 += 1
    elif voto == 3:
        votos_candidato3 += 1
    elif voto == 4:
        votos_candidato4 += 1
    elif voto == 5:
        votos_nulos += 1
    elif voto == 6:
        votos_branco += 1
    else:
        print("Voto inválido.")

print(f'Votos candidato 1: {votos_candidato1}')
print(f'Votos candidato 2: {votos_candidato2}')
print(f'Votos candidato 3: {votos_candidato3}')
print(f'Votos candidato 4: {votos_candidato4}')
print(f'Votos nulos: {votos_nulos}')
print(f'Votos em branco: {votos_branco}')
print(f'Percentual de votos nulos: {(votos_nulos / 20 * 100)}')
print(f'Percentual de votos em branco: {(votos_branco / 20 * 100)}')


## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

In [ ]:
gastos=[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]

media=sum(gastos)/len(gastos)
print(f'{media:.2f}')

3144.22


#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

In [ ]:
gastos=[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]

contador_acima_3k= 0

for i in gastos:
  if i >= 3000:
    contador_acima_3k +=1

porcentagem_acima_3k= 100*contador_acima_3k/len(gastos)

print(f'{contador_acima_3k} compras foram acima de R$ 3000,00 reais. ')
print(f'{porcentagem_acima_3k}% dos gastos foram acima de R$ 3000,00 reais. ')


7 compras foram acima de R$ 3000,00 reais. 
70.0% dos gastos foram acima de R$ 3000,00 reais. 


#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

In [ ]:
five_number=[]

for _ in range(5):
  new_number=int(input('Insira um número: '))
  five_number.append(new_number)

print(five_number)


Insira um número: 5
Insira um número: 1
Insira um número: 45
Insira um número: 8
Insira um número: 7
[5, 1, 45, 8, 7]


#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

In [ ]:
five_number=[]

for _ in range(5):
  new_number=int(input('Insira um número: '))
  five_number.append(new_number)
five_number.reverse()
print(five_number)

Insira um número: 1
Insira um número: 2
Insira um número: 3
Insira um número: 4
Insira um número: 5
[5, 4, 3, 2, 1]


#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

In [ ]:
numb=int(input('Insira um número que queira saber se é primo: '))

prime_list=[]

for i in range(2, numb):
  prime = True

  for div_trial in range (2, i):
    if i % div_trial == 0:
      prime = False #caso haja divisor não é primo
      break

  if prime:
    prime_list.append(i)
print(f'Lista de números primos {prime_list}')


Insira um número que queira saber se é primo: 20
Lista de números primos [2, 3, 5, 7, 11, 13, 17, 19]


#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
